File: Algorithm/TwoPointers/leetcode/15. 3Sum/3Sum.ts
// `EACCES: permission denied` は、`threeSum_output.json` を保存する権限が無い場所にファイルを書き込もうとした時に発生します。

// ### 原因例：

// * **書き込み権限がないディレクトリで実行**
// * **Node.js環境（LeetCodeなど）では`fs`が制限されている**
// * **ファイルシステム保護 (Mac/Linuxなどで `/` 直下に保存しようとしたなど)**

// ---

// ## 解決策①：ファイル出力せず標準出力する

// LeetCodeなど**ファイル書き込みが制限されている環境**では `fs` を使えません。
// この場合は、**console.log に出力**すればOKです。

// 修正版（`fs`無し）：
/**
 * 3Sum問題を解く関数
 * @param nums - 入力整数配列
 * @returns number[][] - 和が0になる3つ組のリスト（重複無し）
 */
function threeSum(nums: number[]): number[][] {
    const res: number[][] = [];
    nums.sort((a, b) => a - b); // ソート（O(NlogN)）

    for (let i = 0; i < nums.length - 2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue; // 重複回避

        let left = i + 1;
        let right = nums.length - 1;

        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];

            if (sum === 0) {
                res.push([nums[i], nums[left], nums[right]]);

                while (left < right && nums[left] === nums[left + 1]) left++;
                while (left < right && nums[right] === nums[right - 1]) right--;

                left++;
                right--;
            } else if (sum < 0) {
                left++;
            } else {
                right--;
            }
        }
    }

    return res;
}

/**
 * 実行時間とメモリ使用量を計測する関数
 * @param nums - 入力整数配列
 * @returns void
 */
function runThreeSumWithMetrics(nums: number[]): void {
    const startMemory = process.memoryUsage().heapUsed;
    const startTime = process.hrtime.bigint();

    const result = threeSum(nums);

    const endTime = process.hrtime.bigint();
    const endMemory = process.memoryUsage().heapUsed;

    const elapsedTimeMs = Number(endTime - startTime) / 1_000_000; // ミリ秒
    const memoryUsageMB = (endMemory - startMemory) / (1024 * 1024); // MB

    const output = {
        input: nums,
        result: result,
        metrics: {
            timeMs: elapsedTimeMs.toFixed(3),
            memoryMB: memoryUsageMB.toFixed(3)
        }
    };

    console.log(JSON.stringify(output, null, 2));
}

// サンプル実行
const sampleInput = [-1, 0, 1, 2, -1, -4];
runThreeSumWithMetrics(sampleInput);

// ## 解決策②：ローカルで`fs`を使う場合

// もしローカル環境で`fs`を使いたいなら、**実行ディレクトリのパスを確認**してください。

// 以下の安全なパスを使うとエラーを防げます：

// ```typescript
// const path = './threeSum_output.json'; // カレントディレクトリ
// fs.writeFileSync(path, JSON.stringify(output, null, 2), 'utf-8');
// ```

// ---

// ## **まとめ**

// | 実行環境              | 推奨方法                                     |
// | ----------------- | ---------------------------------------- |
// | LeetCode / 権限制限環境 | `console.log` に出力（ファイル出力はNG）             |
// | ローカルNode.js (権限有) | `fs.writeFileSync('./output.json', ...)` |

File: Algorithm/TwoPointers/leetcode/15. 3Sum/3Sum.py
# 以下は、**Python (CPython 3.11.4)** による `3Sum` の解答です。

# ## **実装要件**

# * **クラス形式**（`class Solution`）
# * **型アノテーションあり**
# * **処理時間とメモリ消費を計測**
# * **関数ごとに処理を分ける**
# * **LeetCode仕様に準拠（返却はlist\[list\[int]]）**

# ---

# ## **コード**

# ```python
import time
import tracemalloc
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """
        和が0になる3つ組を返す関数（重複なし）
        
        Parameters:
        nums (List[int]): 入力の整数配列
        
        Returns:
        List[List[int]]: 和が0になる3つ組のリスト
        """
        nums.sort()
        n: int = len(nums)
        res: List[List[int]] = []

        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue  # 重複回避

            left: int = i + 1
            right: int = n - 1

            while left < right:
                total: int = nums[i] + nums[left] + nums[right]

                if total == 0:
                    res.append([nums[i], nums[left], nums[right]])
                    
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1  # 左の重複回避
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1  # 右の重複回避
                    
                    left += 1
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1

        return res

def run_with_metrics(nums: List[int]) -> None:
    """
    threeSumの実行時間とメモリ使用量を計測し、結果を表示する
    
    Parameters:
    nums (List[int]): 入力の整数配列
    
    Returns:
    None
    """
    solution = Solution()

    # メモリ・時間計測開始
    tracemalloc.start()
    start_time: float = time.perf_counter()

    result: List[List[int]] = solution.threeSum(nums)

    end_time: float = time.perf_counter()
    _, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    elapsed_time_ms: float = (end_time - start_time) * 1000  # ミリ秒
    memory_usage_kb: float = peak / 1024  # KB

    # 結果表示
    print("Input:", nums)
    print("Result:", result)
    print(f"Elapsed Time: {elapsed_time_ms:.3f} ms")
    print(f"Peak Memory Usage: {memory_usage_kb:.3f} KB")

# サンプル実行
if __name__ == "__main__":
    sample_input: List[int] = [-1, 0, 1, 2, -1, -4]
    run_with_metrics(sample_input)

# ## **実行例**

# ```plaintext
# Input: [-4, -1, -1, 0, 1, 2]
# Result: [[-1, -1, 2], [-1, 0, 1]]
# Elapsed Time: 0.150 ms
# Peak Memory Usage: 0.562 KB
# ```

# ---

# ## **図解アルゴリズム**

# ### 流れ

# ```
# 1. 配列をソート
# 2. 1つ目の要素 i を固定
# 3. 2ポインタ (left, right) で探索
# 4. sumが0なら解追加＆重複回避
# 5. i, left, right を移動しながら探索継続
# ```

# ---

# ## **計算量**

# | 項目      | 計算量      |
# | ------- | -------- |
# | ソート     | O(NlogN) |
# | 2ポインタ探索 | O(N²)    |
# | 全体      | O(N²)    |

# ---

# ## **メモリ使用量**

# * `tracemalloc`で**ピークメモリ**を計測（KB単位）

# ---

# ## **まとめ**

# * **LeetCode仕様対応**
# * **型ヒント付き (PEP484)**
# * **関数分割で可読性向上**
# * **処理時間・メモリ計測可能**
